core/commands/storage.go

```diff
@@ -654,33 +713,62 @@ the chunk and replies back to client for the next challenge step.`,
 		ss := sm.GetOrDefault(ssID)
 		ss.SetStatus(storage.InitStatus)
 		go controlSessionTimeout(ss)
-		chidInt64, err := strconv.ParseInt(channelID, 10, 64)
-		if err != nil {
-			sm.Remove(ssID, "")
-			return err
-		}
-		chID := &ledgerPb.ChannelID{
-			Id: chidInt64,
-		}
 		chunkInfo := ss.GetOrDefault(chunkHash)
-		chunkInfo.UpdateChunk(pid, n.Identity, chID, price)
+		chunkInfo.UpdateChunk(pid, n.Identity, price)
 		chunkInfo.SetState(storage.InitState)
 
-		// build connection with ledger
-		channelInfo, err := getChannelInfo(req.Context, channelID)
-		if err != nil {
-			sendSessionStatusChan(ss.SessionStatusChan, storage.InitStatus, false, err)
-			sm.Remove(ssID, "")
-			return err
-		}
-		log.Debug("Verified channel:", channelInfo)
-
 		sendSessionStatusChan(ss.SessionStatusChan, storage.InitStatus, true, nil)
-		go downloadChunkFromClient(chunkInfo, chunkHash, ssID, n, pid, req, env)
+		// review contract and send back to client
+		go reviewContractAndSign(chunkInfo, chunkHash, ssID, n, pid, req, env, []byte(halfSignedCont))
+		//go downloadChunkFromClient(chunkInfo, chunkHash, ssID, n, pid, req, env)
 		return nil
 	},
 }
 
+func reviewContractAndSign(chunkInfo *storage.Chunk, chunkHash string, ssID string, n *core.IpfsNode,
+	pid peer.ID, req *cmds.Request, env cmds.Environment, payerSig []byte) {
+	// TODO: Check if renter is paid, if so, download file
+	go downloadChunkFromClient(chunkInfo, chunkHash, ssID, n, pid, req, env)
+	// review contract
+	halfSignedContract, err := escrow.UnmarshalEscrowContract(payerSig)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	contract := halfSignedContract.GetContract()
+	// get render's public key
+	payerPubKey, err := pid.ExtractPublicKey()
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	ok, err := crypto.Verify(payerPubKey, contract, halfSignedContract.GetBuyerSignature())
+	if !ok {
+		log.Error("Can't verify contract")
+		return
+	}
+	// Sign on the contract
+	signedContract, err := escrow.SignContractAndMarshal(contract, halfSignedContract, n.PrivateKey, false)
+	if err != nil {
+		log.Error(err)
+		return
+	}
```

Comment:
I personally prefer the "review flow" to be inside core/commands/escrow since it's more escrow related... (not the p2p call though, that is still storage upload command logic)